File: examples/mastra/typescript/tools/queryMemory.ts
import { createTool } from "@mastra/core/tools";
import { z } from "zod";
import { getClient } from "./client.js";
import type { HonchoContext } from "./client.js";

export function makeQueryMemoryTool(ctx: HonchoContext) {
  return createTool({
    id: "query_memory",
    description:
      "Query Honcho's Dialectic API to recall facts about the current user. " +
      "Use this when the user asks what you remember about them.",
    inputSchema: z.object({
      query: z
        .string()
        .describe("Natural language question about the user"),
    }),
    execute: async ({ context: { query } }) => {
      const honcho = getClient();
      const peer = honcho.peer(ctx.userId);
      const response = await peer.chat(query);
      return { result: response ?? "No relevant information found in memory." };
    },
  });
}

File: examples/google-adk/python/tools/query_memory.py
"""Query Honcho memory via the Dialectic API — exposed as a Google ADK tool."""

from .client import get_client


def query_memory(user_id: str, query: str) -> str:
    """Query what Honcho knows about a user using natural language.

    Sends a question to Honcho's Dialectic API and returns an answer grounded
    in the peer's long-term memory and stored observations.

    Args:
        user_id: The user peer ID to query memory for.
        query: Natural language question, e.g. ``"What are my hobbies?"``

    Returns:
        A natural language answer drawn from Honcho's memory, or a fallback
        message if no relevant information was found.

    Raises:
        ValueError: If query is empty or whitespace-only.
        RuntimeError: If the Dialectic API call fails.
    """
    trimmed = query.strip()
    if not trimmed:
        raise ValueError("query must not be empty or whitespace")

    try:
        honcho = get_client()
        peer = honcho.peer(user_id)
        response = peer.chat(query=trimmed)
    except Exception as exc:
        raise RuntimeError("Failed to query Honcho memory") from exc

    if response:
        return str(response)
    return "No relevant information found in memory."
